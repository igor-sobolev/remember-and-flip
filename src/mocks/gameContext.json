{
  "board": [
    { "id": "18", "filled": true, "flip": false },
    { "id": "19", "filled": false, "flip": false },
    { "id": "20", "filled": false, "flip": false },
    { "id": "21", "filled": false, "flip": false },
    { "id": "22", "filled": false, "flip": false },
    { "id": "23", "filled": false, "flip": false },
    { "id": "24", "filled": false, "flip": false },
    { "id": "25", "filled": true, "flip": false },
    { "id": "26", "filled": false, "flip": false },
    { "id": "27", "filled": true, "flip": false },
    { "id": "28", "filled": true, "flip": false },
    { "id": "29", "filled": false, "flip": false },
    { "id": "30", "filled": false, "flip": false },
    { "id": "31", "filled": false, "flip": false },
    { "id": "32", "filled": true, "flip": false },
    { "id": "33", "filled": true, "flip": false }
  ],
  "triesElapsed": 6,
  "gamesElapsed": 1,
  "flipIndex": 0,
  "previewStartTime": 0
}
