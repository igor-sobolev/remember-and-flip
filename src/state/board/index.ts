export { boardMachine } from './board';

export * from './board.types';
