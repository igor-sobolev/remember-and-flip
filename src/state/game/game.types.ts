import { StateNodeConfig } from 'xstate';

export enum GameStates {
  INITIAL = 'initial',
  GAME_INIT = 'gameInit',
  COUNTDOWN = 'countdown',
  GAME = 'game',
  PREVIEW = 'preview',
  RESULT = 'result',
  FINISH = 'finish'
}

export enum GameEvents {
  NEXT = 'next',
  FLIP = 'flip',
  UNFLIP_ALL = 'unflip_all'
}

export enum GameGuards {
  HAS_NO_TRIES_REMAINING = 'hasTriesRemaining',
  HAS_GAMES_REMAINING = 'hasGamesRemaining',
  IS_INITIAL_COUNTDOWN = 'isInitialCountDown',
  HAS_BEEN_PREVIEWED = 'hasBeenPreviewed'
}

export enum GameActions {
  REDUCE_GAMES = 'reduceGames',
  INIT_BOARD = 'initBoard',
  PREVIEW_TILE = 'previewTile',
  CANCEL_PREVIEW_PREVIOUS_TILE = 'cancelPreviewPreviousTile',
  INCREASE_FLIP_INDEX = 'increaseFlipIndex',
  RESET_FLIP_INDEX = 'resetFlipIndex',
  RECORD_PREVIEW_START_TIME = 'recordPreviewStartTime',
  UNFLIP_ALL = 'unflipAll',
  FLIP_EXACT = 'flipExact'
}

export interface GameStateSchema {
  states: {
    [GameStates.INITIAL]: StateNodeConfig<GameContext, GameStateSchema, GameEvent>;
    [GameStates.GAME_INIT]: StateNodeConfig<GameContext, GameStateSchema, GameEvent>;
    [GameStates.COUNTDOWN]: StateNodeConfig<GameContext, GameStateSchema, GameEvent>;
    [GameStates.GAME]: StateNodeConfig<GameContext, GameStateSchema, GameEvent>;
    [GameStates.PREVIEW]: StateNodeConfig<GameContext, GameStateSchema, GameEvent>;
    [GameStates.RESULT]: StateNodeConfig<GameContext, GameStateSchema, GameEvent>;
    [GameStates.FINISH]: StateNodeConfig<GameContext, GameStateSchema, GameEvent>;
  };
}

export type GameEvent =
  | { type: GameEvents.NEXT }
  | { type: GameEvents.UNFLIP_ALL }
  | { type: GameEvents.FLIP; index: number };

export interface GameContext {
  gamesElapsed: number;
  board: Tile[];
  triesElapsed: number;
  flipIndex: number;
  previewStartTime: number;
}

export interface Tile {
  id: string;
  filled: boolean;
  flip: boolean;
}
