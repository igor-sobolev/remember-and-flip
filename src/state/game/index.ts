export { gameMachine } from './game';

export * from './game.types';
