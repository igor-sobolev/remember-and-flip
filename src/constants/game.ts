import { BOARD_SIZE } from './board';

export const GAME_MACHINE_ID = 'board';
export const NUMBER_OF_TRIES = 10;
export const BOARD_COUNTDOWN_DELAY = 3000;
export const GAME_RESULT_DELAY = 2000;
export const BOARD_PREVIEW_DELAY = 10000;
export const GAME_REVEAL_DELAY = 350;
export const TILE_PREVIEW_DELAY = BOARD_PREVIEW_DELAY / Math.pow(BOARD_SIZE, 2);
