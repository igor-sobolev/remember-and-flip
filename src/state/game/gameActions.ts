import { assign, AssignAction } from 'xstate';

import { BOARD_SIZE } from '../../constants';

import { GameContext, GameEvent, Tile } from './game.types';
import { uuid } from '../../components/utils';

export const reduceGames: AssignAction<GameContext, GameEvent> = assign((context: GameContext) => ({
  ...context,
  gamesElapsed: context.gamesElapsed > 0 ? context?.gamesElapsed - 1 : 0
}));

export const initBoard: AssignAction<GameContext, GameEvent> = assign((context: GameContext) => {
  const [board, rightTileNumber] = generateBoard();

  return {
    ...context,
    board,
    triesElapsed: rightTileNumber
  };
});

export const unflipAll: AssignAction<GameContext, GameEvent> = assign((context: GameContext) => {
  const { board } = context;

  return {
    ...context,
    board: board.map((tile: Tile) => ({
      ...tile,
      flip: false
    }))
  };
});

export const flipExact: AssignAction<GameContext, GameEvent> = assign((context: GameContext, event: GameEvent) => {
  const { board } = context;
  const { index } = event as { index: number };

  if (board[index].flip) return context;

  const newTile: Tile = {
    ...board[index],
    flip: true
  };
  const flippedInBoard = [...board.slice(0, index), newTile, ...board.slice(index + 1)];

  return {
    ...context,
    board: flippedInBoard,
    triesElapsed: context.triesElapsed > 0 ? context?.triesElapsed - 1 : 0
  };
});

export const previewTile: AssignAction<GameContext, GameEvent> = assign((context: GameContext) => {
  const { board, flipIndex } = context;

  if (flipIndex === Math.pow(BOARD_SIZE, 2)) return context;

  const newTile: Tile = {
    ...board[flipIndex],
    flip: true
  };
  const flippedInBoard = [...board.slice(0, flipIndex), newTile, ...board.slice(flipIndex + 1)];

  return {
    ...context,
    board: flippedInBoard
  };
});

export const cancelPreviewPreviousTile: AssignAction<GameContext, GameEvent> = assign((context: GameContext) => {
  const { board, flipIndex } = context;

  if (flipIndex === 0) return context;

  const newTile: Tile = {
    ...board[flipIndex - 1],
    flip: false
  };
  const flippedInBoard = [...board.slice(0, flipIndex - 1), newTile, ...board.slice(flipIndex)];

  return {
    ...context,
    board: flippedInBoard
  };
});

export const increaseFlipIndex: AssignAction<GameContext, GameEvent> = assign((context: GameContext) => {
  const { flipIndex } = context;

  return {
    ...context,
    flipIndex: flipIndex < Math.pow(BOARD_SIZE, 2) ? flipIndex + 1 : flipIndex
  };
});

export const resetFlipIndex: AssignAction<GameContext, GameEvent> = assign((context: GameContext) => {
  return {
    ...context,
    flipIndex: 0
  };
});

export const recordPreviewStartTime: AssignAction<GameContext, GameEvent> = assign((context: GameContext) => {
  return {
    ...context,
    previewStartTime: Date.now()
  };
});

function generateBoard(): [Tile[], number] {
  const board: Tile[] = [];
  let filledNumber = 0;

  for (let i = 0; i < BOARD_SIZE; i++) {
    for (let j = 0; j < BOARD_SIZE; j++) {
      const filled = Math.random() < 0.5;
      board.push({
        id: uuid(),
        filled,
        flip: false
      });
      if (filled) filledNumber++;
    }
  }

  return [board, filledNumber];
}
