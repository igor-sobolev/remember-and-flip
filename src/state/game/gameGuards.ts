import { MetaObject } from 'xstate';
import { GameContext, GameEvent, GameStates } from './game.types';

export function hasGamesRemaining(context: GameContext): boolean {
  return context.gamesElapsed !== 0;
}

export function hasTriesRemaining(context: GameContext): boolean {
  return context.triesElapsed !== 0;
}

export function isInitialCountdown(context: GameContext, event: GameEvent, { state }: MetaObject): boolean {
  return !!state.history?.matches(GameStates.GAME_INIT);
}
