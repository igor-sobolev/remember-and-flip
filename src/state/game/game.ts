import { Machine } from 'xstate';

import {
  FINAL_STATE,
  GAME_MACHINE_ID,
  NUMBER_OF_TRIES,
  BOARD_COUNTDOWN_DELAY,
  GAME_RESULT_DELAY,
  BOARD_PREVIEW_DELAY,
  TILE_PREVIEW_DELAY
} from '../../constants';

import { GameStates, GameStateSchema, GameEvent, GameContext, GameEvents, GameActions, GameGuards } from './game.types';
import { hasNoTriesRemaining, isInitialCountdown, hasGamesRemaining, hasBeenPreviewed } from './gameGuards';
import {
  reduceGames,
  initBoard,
  previewTile,
  cancelPreviewPreviousTile,
  increaseFlipIndex,
  resetFlipIndex,
  recordPreviewStartTime,
  unflipAll,
  flipExact
} from './gameActions';

const conditionalTransition = (target: string, cond: string, actions: string[] = []) => ({
  target,
  cond,
  actions
});

const directTransition = (target: string, actions: string[] = []) => ({ target, actions });

export const gameMachine = Machine<GameContext, GameStateSchema, GameEvent>(
  {
    key: GAME_MACHINE_ID,
    initial: GameStates.INITIAL,
    context: {
      gamesElapsed: NUMBER_OF_TRIES,
      board: [],
      triesElapsed: 0,
      flipIndex: 0,
      previewStartTime: 0
    },
    states: {
      [GameStates.INITIAL]: {
        on: {
          [GameEvents.NEXT]: directTransition(GameStates.GAME_INIT)
        }
      },
      [GameStates.GAME_INIT]: {
        entry: [GameActions.INIT_BOARD],
        after: {
          IMMEDIATE: directTransition(GameStates.COUNTDOWN)
        }
      },
      [GameStates.COUNTDOWN]: {
        after: {
          BOARD_COUNTDOWN_DELAY: [
            conditionalTransition(GameStates.PREVIEW, GameGuards.IS_INITIAL_COUNTDOWN, [
              GameActions.RECORD_PREVIEW_START_TIME
            ]),
            directTransition(GameStates.GAME)
          ]
        },
        exit: [GameActions.RESET_FLIP_INDEX]
      },
      [GameStates.GAME]: {
        after: {
          IMMEDIATE: conditionalTransition(GameStates.RESULT, GameGuards.HAS_NO_TRIES_REMAINING)
        },
        on: {
          [GameEvents.FLIP]: directTransition(GameStates.GAME, [GameActions.FLIP_EXACT])
        }
      },
      [GameStates.PREVIEW]: {
        entry: [GameActions.PREVIEW_TILE, GameActions.CANCEL_PREVIEW_PREVIOUS_TILE],
        after: {
          TILE_PREVIEW_DELAY: [
            conditionalTransition(GameStates.COUNTDOWN, GameGuards.HAS_BEEN_PREVIEWED),
            directTransition(GameStates.PREVIEW, [GameActions.INCREASE_FLIP_INDEX])
          ]
        }
      },
      [GameStates.RESULT]: {
        exit: [GameActions.REDUCE_GAMES],
        after: {
          UNFLIP_DELAY: {
            actions: [GameActions.UNFLIP_ALL]
          },
          GAME_RESULT_DELAY: [
            conditionalTransition(GameStates.GAME_INIT, GameGuards.HAS_GAMES_REMAINING),
            directTransition(GameStates.FINISH)
          ]
        }
      },
      [GameStates.FINISH]: {
        type: FINAL_STATE
      }
    }
  },
  {
    guards: {
      [GameGuards.HAS_GAMES_REMAINING]: hasGamesRemaining,
      [GameGuards.HAS_NO_TRIES_REMAINING]: hasNoTriesRemaining,
      [GameGuards.IS_INITIAL_COUNTDOWN]: isInitialCountdown,
      [GameGuards.HAS_BEEN_PREVIEWED]: hasBeenPreviewed
    },
    actions: {
      [GameActions.REDUCE_GAMES]: reduceGames,
      [GameActions.INIT_BOARD]: initBoard,
      [GameActions.PREVIEW_TILE]: previewTile,
      [GameActions.CANCEL_PREVIEW_PREVIOUS_TILE]: cancelPreviewPreviousTile,
      [GameActions.INCREASE_FLIP_INDEX]: increaseFlipIndex,
      [GameActions.RESET_FLIP_INDEX]: resetFlipIndex,
      [GameActions.RECORD_PREVIEW_START_TIME]: recordPreviewStartTime,
      [GameActions.FLIP_EXACT]: flipExact,
      [GameActions.UNFLIP_ALL]: unflipAll
    },
    delays: {
      BOARD_COUNTDOWN_DELAY,
      GAME_RESULT_DELAY,
      UNFLIP_DELAY: GAME_RESULT_DELAY / 2,
      BOARD_PREVIEW_DELAY,
      TILE_PREVIEW_DELAY,
      IMMEDIATE: 0
    }
  }
);
