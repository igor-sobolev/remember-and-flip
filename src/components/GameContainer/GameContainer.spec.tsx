import * as React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import { GameContainer } from '.';
import { useMachine } from '@xstate/react';
import { GameStates } from '../../state/game';
import * as gameContextMock from '../../mocks/gameContext.json';

const sendMock = jest.fn();

jest.mock('@xstate/react', () => ({
  useMachine: jest
    .fn()
    .mockImplementation(() => [{ value: GameStates.INITIAL, context: { ...gameContextMock, board: [] } }, sendMock])
}));

describe('given GameContainer component', () => {
  let wrapper: ShallowWrapper;

  beforeEach(() => {
    wrapper = shallow(<GameContainer />);
  });

  afterAll(() => {
    jest.resetAllMocks();
  });

  it('should match the snapshot', () => {
    expect(wrapper.getElement()).toMatchSnapshot();
  });

  describe('when machine returns GAME state', () => {
    beforeEach(() => {
      (useMachine as jest.Mock).mockImplementation(() => [
        { value: GameStates.GAME, context: gameContextMock },
        sendMock
      ]);

      wrapper = shallow(<GameContainer />);
    });

    it('should match the snapshot', () => {
      expect(wrapper.getElement()).toMatchSnapshot();
    });
  });

  describe('when machine returns RESULT state', () => {
    beforeEach(() => {
      (useMachine as jest.Mock).mockImplementation(() => [
        { value: GameStates.RESULT, context: gameContextMock },
        sendMock
      ]);

      wrapper = shallow(<GameContainer />);
    });

    it('should match the snapshot', () => {
      expect(wrapper.getElement()).toMatchSnapshot();
    });
  });

  describe('when machine returns FINISH state', () => {
    beforeEach(() => {
      (useMachine as jest.Mock).mockImplementation(() => [
        { value: GameStates.FINISH, context: gameContextMock },
        sendMock
      ]);

      wrapper = shallow(<GameContainer />);
    });

    it('should match the snapshot', () => {
      expect(wrapper.getElement()).toMatchSnapshot();
    });
  });

  describe('when machine returns COUNTDOWN state', () => {
    beforeEach(() => {
      (useMachine as jest.Mock).mockImplementation(() => [
        { value: GameStates.COUNTDOWN, context: gameContextMock },
        sendMock
      ]);

      wrapper = shallow(<GameContainer />);
    });

    it('should match the snapshot', () => {
      expect(wrapper.getElement()).toMatchSnapshot();
    });
  });

  describe('when machine returns PREVIEW state', () => {
    beforeEach(() => {
      (useMachine as jest.Mock).mockImplementation(() => [
        { value: GameStates.PREVIEW, context: gameContextMock },
        sendMock
      ]);

      wrapper = shallow(<GameContainer />);
    });

    it('should match the snapshot', () => {
      expect(wrapper.getElement()).toMatchSnapshot();
    });
  });

  describe('when machine returns GAME_INIT state', () => {
    beforeEach(() => {
      (useMachine as jest.Mock).mockImplementation(() => [
        { value: GameStates.GAME_INIT, context: gameContextMock },
        sendMock
      ]);

      wrapper = shallow(<GameContainer />);
    });

    it('should match the snapshot', () => {
      expect(wrapper.getElement()).toMatchSnapshot();
    });
  });
});
