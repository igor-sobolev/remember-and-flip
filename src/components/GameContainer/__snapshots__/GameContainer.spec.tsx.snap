// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`given GameContainer component should match the snapshot 1`] = `
<div
  className="game"
>
  <React.Fragment>
    <Board
      board={Array []}
      boardWidth={614.4000000000001}
      disabled={true}
    />
    <ModalInfo>
      <Button
        data-qa="GameContainer_next"
        onClick={[Function]}
      >
        Start
      </Button>
    </ModalInfo>
  </React.Fragment>
</div>
`;

exports[`given GameContainer component when machine returns COUNTDOWN state should match the snapshot 1`] = `
<div
  className="game"
>
  <React.Fragment>
    <Board
      board={
        Array [
          Object {
            "filled": true,
            "flip": false,
            "id": "18",
          },
          Object {
            "filled": false,
            "flip": false,
            "id": "19",
          },
          Object {
            "filled": false,
            "flip": false,
            "id": "20",
          },
          Object {
            "filled": false,
            "flip": false,
            "id": "21",
          },
          Object {
            "filled": false,
            "flip": false,
            "id": "22",
          },
          Object {
            "filled": false,
            "flip": false,
            "id": "23",
          },
          Object {
            "filled": false,
            "flip": false,
            "id": "24",
          },
          Object {
            "filled": true,
            "flip": false,
            "id": "25",
          },
          Object {
            "filled": false,
            "flip": false,
            "id": "26",
          },
          Object {
            "filled": true,
            "flip": false,
            "id": "27",
          },
          Object {
            "filled": true,
            "flip": false,
            "id": "28",
          },
          Object {
            "filled": false,
            "flip": false,
            "id": "29",
          },
          Object {
            "filled": false,
            "flip": false,
            "id": "30",
          },
          Object {
            "filled": false,
            "flip": false,
            "id": "31",
          },
          Object {
            "filled": true,
            "flip": false,
            "id": "32",
          },
          Object {
            "filled": true,
            "flip": false,
            "id": "33",
          },
        ]
      }
      boardWidth={0}
      disabled={true}
    />
    <ModalInfo>
      Ready!
    </ModalInfo>
  </React.Fragment>
</div>
`;

exports[`given GameContainer component when machine returns FINISH state should match the snapshot 1`] = `
<div
  className="game"
>
  <React.Fragment>
    <Board
      board={
        Array [
          Object {
            "filled": true,
            "flip": false,
            "id": "18",
          },
          Object {
            "filled": false,
            "flip": false,
            "id": "19",
          },
          Object {
            "filled": false,
            "flip": false,
            "id": "20",
          },
          Object {
            "filled": false,
            "flip": false,
            "id": "21",
          },
          Object {
            "filled": false,
            "flip": false,
            "id": "22",
          },
          Object {
            "filled": false,
            "flip": false,
            "id": "23",
          },
          Object {
            "filled": false,
            "flip": false,
            "id": "24",
          },
          Object {
            "filled": true,
            "flip": false,
            "id": "25",
          },
          Object {
            "filled": false,
            "flip": false,
            "id": "26",
          },
          Object {
            "filled": true,
            "flip": false,
            "id": "27",
          },
          Object {
            "filled": true,
            "flip": false,
            "id": "28",
          },
          Object {
            "filled": false,
            "flip": false,
            "id": "29",
          },
          Object {
            "filled": false,
            "flip": false,
            "id": "30",
          },
          Object {
            "filled": false,
            "flip": false,
            "id": "31",
          },
          Object {
            "filled": true,
            "flip": false,
            "id": "32",
          },
          Object {
            "filled": true,
            "flip": false,
            "id": "33",
          },
        ]
      }
      boardWidth={0}
      disabled={true}
    />
    <ModalInfo>
      Congratulations!
    </ModalInfo>
  </React.Fragment>
</div>
`;

exports[`given GameContainer component when machine returns GAME state should match the snapshot 1`] = `
<div
  className="game"
>
  <Board
    board={
      Array [
        Object {
          "filled": true,
          "flip": false,
          "id": "18",
        },
        Object {
          "filled": false,
          "flip": false,
          "id": "19",
        },
        Object {
          "filled": false,
          "flip": false,
          "id": "20",
        },
        Object {
          "filled": false,
          "flip": false,
          "id": "21",
        },
        Object {
          "filled": false,
          "flip": false,
          "id": "22",
        },
        Object {
          "filled": false,
          "flip": false,
          "id": "23",
        },
        Object {
          "filled": false,
          "flip": false,
          "id": "24",
        },
        Object {
          "filled": true,
          "flip": false,
          "id": "25",
        },
        Object {
          "filled": false,
          "flip": false,
          "id": "26",
        },
        Object {
          "filled": true,
          "flip": false,
          "id": "27",
        },
        Object {
          "filled": true,
          "flip": false,
          "id": "28",
        },
        Object {
          "filled": false,
          "flip": false,
          "id": "29",
        },
        Object {
          "filled": false,
          "flip": false,
          "id": "30",
        },
        Object {
          "filled": false,
          "flip": false,
          "id": "31",
        },
        Object {
          "filled": true,
          "flip": false,
          "id": "32",
        },
        Object {
          "filled": true,
          "flip": false,
          "id": "33",
        },
      ]
    }
    boardWidth={0}
    onFlip={[Function]}
  />
</div>
`;

exports[`given GameContainer component when machine returns GAME_INIT state should match the snapshot 1`] = `
<div
  className="game"
>
  <React.Fragment>
    <Board
      board={
        Array [
          Object {
            "filled": true,
            "flip": false,
            "id": "18",
          },
          Object {
            "filled": false,
            "flip": false,
            "id": "19",
          },
          Object {
            "filled": false,
            "flip": false,
            "id": "20",
          },
          Object {
            "filled": false,
            "flip": false,
            "id": "21",
          },
          Object {
            "filled": false,
            "flip": false,
            "id": "22",
          },
          Object {
            "filled": false,
            "flip": false,
            "id": "23",
          },
          Object {
            "filled": false,
            "flip": false,
            "id": "24",
          },
          Object {
            "filled": true,
            "flip": false,
            "id": "25",
          },
          Object {
            "filled": false,
            "flip": false,
            "id": "26",
          },
          Object {
            "filled": true,
            "flip": false,
            "id": "27",
          },
          Object {
            "filled": true,
            "flip": false,
            "id": "28",
          },
          Object {
            "filled": false,
            "flip": false,
            "id": "29",
          },
          Object {
            "filled": false,
            "flip": false,
            "id": "30",
          },
          Object {
            "filled": false,
            "flip": false,
            "id": "31",
          },
          Object {
            "filled": true,
            "flip": false,
            "id": "32",
          },
          Object {
            "filled": true,
            "flip": false,
            "id": "33",
          },
        ]
      }
      boardWidth={0}
      disabled={true}
    />
    <ModalInfo>
      <Spinner />
    </ModalInfo>
  </React.Fragment>
</div>
`;

exports[`given GameContainer component when machine returns PREVIEW state should match the snapshot 1`] = `
<div
  className="game"
>
  <React.Fragment>
    <Board
      board={
        Array [
          Object {
            "filled": true,
            "flip": false,
            "id": "18",
          },
          Object {
            "filled": false,
            "flip": false,
            "id": "19",
          },
          Object {
            "filled": false,
            "flip": false,
            "id": "20",
          },
          Object {
            "filled": false,
            "flip": false,
            "id": "21",
          },
          Object {
            "filled": false,
            "flip": false,
            "id": "22",
          },
          Object {
            "filled": false,
            "flip": false,
            "id": "23",
          },
          Object {
            "filled": false,
            "flip": false,
            "id": "24",
          },
          Object {
            "filled": true,
            "flip": false,
            "id": "25",
          },
          Object {
            "filled": false,
            "flip": false,
            "id": "26",
          },
          Object {
            "filled": true,
            "flip": false,
            "id": "27",
          },
          Object {
            "filled": true,
            "flip": false,
            "id": "28",
          },
          Object {
            "filled": false,
            "flip": false,
            "id": "29",
          },
          Object {
            "filled": false,
            "flip": false,
            "id": "30",
          },
          Object {
            "filled": false,
            "flip": false,
            "id": "31",
          },
          Object {
            "filled": true,
            "flip": false,
            "id": "32",
          },
          Object {
            "filled": true,
            "flip": false,
            "id": "33",
          },
        ]
      }
      boardWidth={0}
      disabled={true}
    />
    <ModalInfo>
      <Countdown$1
        autoStart={true}
        controlled={false}
        date={10000}
        daysInHours={false}
        intervalDelay={1000}
        precision={0}
        renderer={[Function]}
        zeroPadTime={2}
      />
    </ModalInfo>
  </React.Fragment>
</div>
`;

exports[`given GameContainer component when machine returns RESULT state should match the snapshot 1`] = `
<div
  className="game"
>
  <React.Fragment>
    <Board
      board={
        Array [
          Object {
            "filled": true,
            "flip": false,
            "id": "18",
          },
          Object {
            "filled": false,
            "flip": false,
            "id": "19",
          },
          Object {
            "filled": false,
            "flip": false,
            "id": "20",
          },
          Object {
            "filled": false,
            "flip": false,
            "id": "21",
          },
          Object {
            "filled": false,
            "flip": false,
            "id": "22",
          },
          Object {
            "filled": false,
            "flip": false,
            "id": "23",
          },
          Object {
            "filled": false,
            "flip": false,
            "id": "24",
          },
          Object {
            "filled": true,
            "flip": false,
            "id": "25",
          },
          Object {
            "filled": false,
            "flip": false,
            "id": "26",
          },
          Object {
            "filled": true,
            "flip": false,
            "id": "27",
          },
          Object {
            "filled": true,
            "flip": false,
            "id": "28",
          },
          Object {
            "filled": false,
            "flip": false,
            "id": "29",
          },
          Object {
            "filled": false,
            "flip": false,
            "id": "30",
          },
          Object {
            "filled": false,
            "flip": false,
            "id": "31",
          },
          Object {
            "filled": true,
            "flip": false,
            "id": "32",
          },
          Object {
            "filled": true,
            "flip": false,
            "id": "33",
          },
        ]
      }
      boardWidth={0}
      disabled={true}
    />
    <ModalInfo>
      Great!
    </ModalInfo>
  </React.Fragment>
</div>
`;
